components/console_cmd_mqtt/console_mqtt.c

```diff
@@ -0,0 +1,445 @@
+            ESP_LOGI(TAG, "Last errno string (%s)", strerror(event->error_handle->esp_transport_sock_errno));
+        }
+        break;
+    default:
+        ESP_LOGI(TAG, "Other event id:%d", event->event_id);
+        break;
+    }
+}
+
+
+static esp_mqtt_client_config_t
+get_mqtt_config_basic(const char *uri_i)
+{
+    char *uri;
+    asprintf(&uri, "%s", uri_i);
+
+    esp_mqtt_client_config_t mqtt_cfg = {
+        .broker.address.uri = uri,
+    };
+
+    return mqtt_cfg;
+}
+
+
+static esp_mqtt_client_config_t
+get_mqtt_config_username_password(const char *uri_i, const char *username_i, const char *password_i)
+{
+    char *uri, *username, *password;
+
+    asprintf(&uri, "%s", uri_i);
+    asprintf(&username, "%s", username_i);
+    asprintf(&password, "%s", password_i);
+
+    esp_mqtt_client_config_t mqtt_cfg = {
+        .broker.address.uri = uri,
+        .credentials.username = username,
+        .credentials.authentication.password = password
+    };
+
+    return mqtt_cfg;
+}
+
+
+static void mqtt_config_free(esp_mqtt_client_config_t mqtt_cfg)
+{
+    if (mqtt_cfg.broker.address.uri) {
+        free((char *)mqtt_cfg.broker.address.uri);
+    }
+
+    if (mqtt_cfg.credentials.username) {
+        free((char *)mqtt_cfg.credentials.username);
+    }
+
+    if (mqtt_cfg.credentials.authentication.password) {
+        free((char *)mqtt_cfg.credentials.authentication.password);
+    }
+}
+
+
+static int do_mqtt_cmd(int argc, char **argv)
+{
+    int nerrors = arg_parse(argc, argv, (void **)&mqtt_args);
+    if (nerrors != 0) {
+        arg_print_errors(stderr, mqtt_args.end, argv[0]);
+        return 1;
+    }
+
+    if (mqtt_args.connect->count > 0) {
+        esp_mqtt_client_config_t mqtt_cfg;
+
+        if (client_handle != NULL) {
+            ESP_LOGI(TAG, "mqtt client already connected");
+            return 0;
+        }
+
+        if ((mqtt_args.username->count > 0) && (mqtt_args.password->count > 0)) {
+            if (mqtt_args.uri->count > 0) {
+                mqtt_cfg = get_mqtt_config_username_password(mqtt_args.uri->sval[0],
+                           mqtt_args.username->sval[0],
+                           mqtt_args.password->sval[0]);
+            } else {
+                mqtt_cfg = get_mqtt_config_username_password(CONFIG_MQTT_BROKER_URL,
+                           mqtt_args.username->sval[0],
+                           mqtt_args.password->sval[0]);
+            }
+        } else {
+            if (mqtt_args.uri->count > 0) {
+                mqtt_cfg = get_mqtt_config_basic(mqtt_args.uri->sval[0]);
+            } else {
+                mqtt_cfg = get_mqtt_config_basic(CONFIG_MQTT_BROKER_URL);
+            }
+        }
+
+        ESP_LOGI(TAG, "broker: %s", mqtt_cfg.broker.address.uri);
+
+        if (mqtt_args.cafile->count > 0) {
+            if (server_cert_pem_start && server_cert_pem_end) {
+                mqtt_cfg.broker.verification.certificate = (const char *)server_cert_pem_start;
+            }
```

Comment:
maybe we could provide an error message, as the user added `--cafile` param, but didn't add the actual file?

Reply:
same for client's key and cert

Reply:
Add a log saying the file is not supplied.